drivers/ethernet/eth_stm32_hal.c

```diff
@@ -221,18 +221,32 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
 	tx_config.Length = total_len;
 	tx_config.TxBuffer = tx_buffer_def;
 
-	hal_ret = HAL_ETH_Transmit(heth, &tx_config, ETH_DMA_TRANSMIT_TIMEOUT);
+	/* tx_buffer is allocated on function stack, we need to wait for the transfer to complete */
+	/* So it is not freed before the interrupt happens */
+	hal_ret = HAL_ETH_Transmit_IT(heth, &tx_config);
+
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("HAL_ETH_Transmit: failed!");
+		res = -EIO;
+		goto error;
+	}
+
+	/* Wait for end of TX buffer transmission */
+	if (k_sem_take(&dev_data->tx_int_sem, K_FOREVER) != 0) {
```

Comment:
> Hello - I'm thinking about the `K_FOREVER` here. While testing this commit with Civetweb, I'm definitely able to get into situation when I have a deadlock. I still do not know what is the reason, I have a feeling that civetweb do not play well with current TCP stack.
> I tried to change it to some long timeout (20s) and at least it prevents (or more precisely give a chance to recover from) the deadlocks...

Are you using civetweb http-server?

Reply:
> Are you using civetweb http-server?
Yes, I am. I have a simple site that provides a web interface to a sensor. There is an ajax routine that refreshes the information each 2s by request to a page returning short json string. When I wait long enough I usually see a deadlock here. I have not find exact steps to reproduce it though...

Reply:
> > Are you using civetweb http-server?
> 
> Yes, I am. I have a simple site that provides a web interface to a sensor. There is an ajax routine that refreshes the information each 2s by request to a page returning short json string. When I wait long enough I usually see a deadlock here. I have not find exact steps to reproduce it though...

Does the master's branch version of the driver work?

Reply:
> Does the master's branch version of the driver work?

Hmmm, I need to check. But what do you mean by *master's branch version of the driver*? I think I have merged `lochej/drivers/stm32h7_it_based_eth` into my master, but it was probably very close.

Reply:
> > Does the master's branch version of the driver work?
> 
> Hmmm, I need to check. But what do you mean by *master's branch version of the driver*? I think I have merged `lochej/drivers/stm32h7_it_based_eth` into my master, but it was probably very close.

I mean [this one](https://github.com/zephyrproject-rtos/zephyr.git). It has currently h7 eth driver support implemented as close as possible to F7 one.

Reply:
I might have a solution to avoid having K_FOREVER here and keep a timeout. 

If you don't get the a TX complete interrupt that unlocks the semaphore, that means there must be some error that happened during DMA transmission or eth interrupt. I'll investigate on this issue but I was thinking about adding a timeout and semaphore reset to handle DMA transmission errors.

I'll be able to test later in the week but I'll update the PR so @Nukersson and @xhpohanka can test the error handling.

Reply:
> I might have a solution to avoid having K_FOREVER here and keep a timeout. 
> 
> If you don't get the a TX complete interrupt that unlocks the semaphore, that means there must be some error that happened during DMA transmission or eth interrupt. I'll investigate on this issue but I was thinking about adding a timeout and semaphore reset to handle DMA transmission errors.
> 
> I'll be able to test later in the week but I'll update the PR so @Nukersson and @xhpohanka can test the error handling.

Good idea. There should be DMA/ETH hardware error handlers in stm32h7 driver implementation. This would allow us handle described errors and/or drop/handle corrupted packets. I can take a look at posaible solution next week.

Reply:
@Nukersson @xhpohanka I've rolled out modifications to the driver to add the timeout to the semaphore and error logging. You can give it a try !

Reply:
Thank you for the update @lochej. I do not have time to analyze it more today but FYI I still got `<err> eth_stm32_hal: HAL_ETH_TransmitIT tx_int_sem take timeout` quite often. Neither ETH nor DMA error is signaled.

Reply:
@lochej here is a log of  timeouts,  I will probably share some more later, but now I changed settings a bit and I cannot reproduce it so easy like last week
```
[00:00:05.049,000] <err> eth_stm32_hal: HAL_ETH_TransmitIT tx_int_sem take timeout
[00:00:05.049,000] <err> eth_stm32_hal: eth packet timeout
d8 d0 90 52 8b 2e cc bd  35 ff ff ff 08 00 45 00 |...R.... 5.....E.
00 28 00 00 00 00 40 06  f5 c6 c0 a8 01 fb c0 a8 |.(....@. ........
01 be 93 5b 07 5b 2c 6f  c4 9a e5 b2 a1 af 50 10 |...[.[,o ......P.
05 00 12 a8 00 00                                |......           
[00:00:05.247,000] <inf> mqtt: MQTT client connected!
[00:00:05.264,000] <err> eth_stm32_hal: HAL_ETH_TransmitIT tx_int_sem take timeout
[00:00:05.265,000] <err> eth_stm32_hal: eth packet timeout
d8 d0 90 52 8b 2e cc bd  35 ff ff ff 08 00 45 00 |...R.... 5.....E.
00 28 00 00 00 00 40 06  f5 c6 c0 a8 01 fb c0 a8 |.(....@. ........
01 be 93 5b 07 5b 2c 6f  c4 b4 e5 b2 a1 b3 50 10 |...[.[,o ......P.
05 00 12 8a 00 00                                |......           
[00:00:05.917,000] <err> eth_stm32_hal: HAL_ETH_TransmitIT tx_int_sem take timeout
[00:00:05.917,000] <err> eth_stm32_hal: eth packet timeout
d8 d0 90 52 8b 2e cc bd  35 ff ff ff 08 00 45 00 |...R.... 5.....E.
00 44 00 00 00 00 40 01  f5 af c0 a8 01 fb c0 a8 |.D....@. ........
01 be 03 03 82 21 00 00  00 00 45 00 00 28 ed 8a |.....!.. ..E..(..
40 00 40 06 c8 3b c0 a8  01 be c0 a8 01 fb 07 5b |@.@..;.. .......[
80 0a 1c ce be dc 24 d8  1d 81 50 11 fa 1c 8b 43 |......$. ..P....C
00 00                                            |..               
[00:00:06.767,000] <inf> mqtt: PUBACK packet id: 56264
[00:00:08.283,000] <inf> mqtt: PUBACK packet id: 51096
[00:00:08.475,000] <err> eth_stm32_hal: HAL_ETH_TransmitIT tx_int_sem take timeout
[00:00:08.476,000] <err> eth_stm32_hal: eth packet timeout
d8 d0 90 52 8b 2e cc bd  35 ff ff ff 08 00 45 00 |...R.... 5.....E.
00 44 00 00 00 00 40 01  f5 af c0 a8 01 fb c0 a8 |.D....@. ........
01 be 03 03 82 21 00 00  00 00 45 00 00 28 ed 8c |.....!.. ..E..(..
40 00 40 06 c8 39 c0 a8  01 be c0 a8 01 fb 07 5b |@.@..9.. .......[
80 0a 1c ce be dc 24 d8  1d 81 50 11 fa 1c 8b 43 |......$. ..P....C
00 00                                            |..   
```